debian/noiz2sa.6
.TH NOIZ2SA 6 "15.11.2004" "abstract arcade shooter"

.SH NAME
noiz2sa \- an abstract arcade shooter

.SH SYNOPSIS
.B noiz2sa [\fI-nosound\fR] [\fI-window\fR] [\fI-reverse\fR] [\fI-brightness
n\fR] [\fI-nowait\fR] [\fI-accframe\fR]

.SH DESCRIPTION
noiz2sa is an abstract arcade shooter, just fire it up and you'll find out
what that means.

.SH USAGE
You can play the game with a keyboard or a joystick. On the keyboard the following keys are used:

 - Movement   Arrow keys
 - Fire       [Z]       
 - Slowdown   [X]      
 - Pause      [P]
 - Leave Game [Esc]

Control your ship and avoid the barrage.
A ship is not destroyed even if it contacts an enemy main body.
A ship becomes slow while holding the slowdown key.

A green star is the bonus item.
A score of the item(displayed at the left-up corner) increases 
if you get items continuously.

When all ships are destroyed, the game is over.
The ship extends 200,000 and every 500,000 points.

Apart from that the game is really easy, no need for a manual!

.SH OPTIONS

These command line options are available:
 \-nosound       Stop the sound.
 \-window        Launch the game in the window, not use the full-screen.
 \-reverse       Reverse the fire key and the slowdown key.
 \-brightness n  Set the brightness of the sceen(n=0-256).
 \-accframe      Use the alternative framerate management algorithm.
                (If you have a problem with framerate, try this option.)


.SH AUTHOR
This manual page was written for the Debian GNU/Linux system by Robert Lemmen
<robertle@semistable.com> (but may be used by others, of course)


debian/changelog
noiz2sa (0.51a-10) unstable; urgency=medium

  * Maintenance Release
    * Added debhelper misc-depends
    * Added recommended default targets to debian/rules
    * Use dh_prep instead of dh_clean
    * Switch to debhelper compat level 7
    * Removed debian/watch
    * Added debian/source/format
    * Update debian/rules with some hardening flags
    * Set standards-version to 3.9.8

 -- Robert Lemmen <robertle@semistable.com>  Sat, 11 Jun 2016 08:43:55 +0100

noiz2sa (0.51a-9) unstable; urgency=low

  * Link with -lm (closes: #555774)
  * Fixed some conversion errors (closes: #646453)
  * Added a .desktop files (closes: #538036)
  * Update to newer policy version

 -- Robert Lemmen <robertle@semistable.com>  Wed, 16 Nov 2011 17:58:48 +0000

noiz2sa (0.51a-8) unstable; urgency=low

  * More fixes for GCC 4.3 (closes: #455140)
  * Fixed the uscan watch file (closes: #450021)

 -- Robert Lemmen <robertle@semistable.com>  Wed,  5 Mar 2008 17:50:27 +0000

noiz2sa (0.51a-7) unstable; urgency=low

  * Added a menu item

 -- Robert Lemmen <robertle@semistable.com>  Tue, 28 Aug 2007 18:27:01 +0200

noiz2sa (0.51a-6) unstable; urgency=low

  * Include additional headers to make noiz2sa build with GCC 4.3 
    (closes: #417450)
  * Made the package binNMU-friendly

 -- Robert Lemmen <robertle@semistable.com>  Mon, 16 Apr 2007 10:04:30 +0200

noiz2sa (0.51a-5) unstable; urgency=low

  * Fixed a problem that made the build fail with g++ 4.1, thanks tbm
    (closes: #356424)

 -- Robert Lemmen <robertle@semistable.com>  Tue, 21 Mar 2006 09:49:26 +0100

noiz2sa (0.51a-4) unstable; urgency=low

  * Fixed the debian/rules file to properly pass compiler options to the
    makefiles (also fixes FTBFS problems on m68k)

 -- Robert Lemmen <robertle@semistable.com>  Fri, 19 Aug 2005 10:29:18 +0200

noiz2sa (0.51a-3) unstable; urgency=low

  * Rebuild for g++ transition
  * Updated to newer standards version

 -- Robert Lemmen <robertle@semistable.com>  Tue, 19 Jul 2005 10:56:44 +0200

noiz2sa (0.51a-2) unstable; urgency=low

  * Uses the libboost package now instead of it's own copy of the files
  * Makefile cleanups
  * Also clean tinyxml subdir

 -- Robert Lemmen <robertle@semistable.com>  Mon, 22 Nov 2004 19:26:48 +0100

noiz2sa (0.51a-1) unstable; urgency=low

  * Initial Release. (closes: #259066)

 -- Robert Lemmen <robertle@semistable.com>  Mon, 15 Nov 2004 11:37:18 +0100



src/noiz2sa.c
/*
 * $Id: noiz2sa.c,v 1.8 2003/02/12 13:55:13 kenta Exp $
 *
 * Copyright 2002 Kenta Cho. All rights reserved.
 */

/**
 * Noiz2sa main routine.
 *
 * @version $Revision: 1.8 $
 */
#include "SDL.h"
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#include "noiz2sa.h"
#include "screen.h"
#include "vector.h"
#include "ship.h"
#include "shot.h"
#include "frag.h"
#include "bonus.h"
#include "foe_mtd.h"
#include "brgmng_mtd.h"
#include "background.h"
#include "degutil.h"
#include "soundmanager.h"
#include "attractmanager.h"
#include "gamepad.h"

static int noSound = 0;

// Initialize and load preference.
static void initFirst() {
  loadPreference();
  srand(SDL_GetTicks());
  initBarragemanager();
  initAttractManager();
}

// Quit and save preference.
void quitLast() {
  if ( !noSound ) closeSound();
  savePreference();
  closeBarragemanager();
  closeSDL();
  SDL_Quit();
  exit(1);
}

int status;

static float stagePrm[STAGE_NUM+ENDLESS_STAGE_NUM+1][3] = {
  {13, 0.5f, 0.12f}, {2, 1.8f, 0.15f}, {3, 3.2f, 0.1f}, {90, 6.0f, 0.3f}, {5, 5.0f, 0.6f},
  {6, 10.0f, 0.6f}, {7, 5.0f, 2.2f}, {98, 12.0f, 1.5f}, {9, 10.0f, 2.0f}, {79, 21.0f, 1.5f},
  {-3, 5.0f, 0.7f}, {-1, 10.0f, 1.2f}, {-4, 15.0f, 1.8f}, {-2, 16.0f, 1.8f},
  {0, -1.0f, 0.0f},
};

void initTitleStage(int stg) {
  initFoes();
  initBarrages(stagePrm[stg][0], stagePrm[stg][1], stagePrm[stg][2]);
}

void initTitle() {
  int stg;
  status = TITLE;

  stg = initTitleAtr();
  initShip();
  initShots();
  initFrags();
  initBonuses();
  initBackground();
  setStageBackground(1);

  initTitleStage(stg);
}

void initGame(int stg) {
  status = IN_GAME;

  initShip();
  initShots();
  initFoes();
  initFrags();
  initBonuses();
  initBackground();

  initBarrages(stagePrm[stg][0], stagePrm[stg][1], stagePrm[stg][2]);
  initGameState(stg);
  if ( stg < STAGE_NUM ) {
    setStageBackground(stg%5+1);
    playMusic(stg%5+1);
  } else {
    if ( !insane ) {
      setStageBackground(0);
      playMusic(0);
    } else {
      setStageBackground(6);
      playMusic(6);
    }
  }
}

void initGameover() {
  status = GAMEOVER;
  initGameoverAtr();
}

void initStageClear() {
  status = STAGE_CLEAR;
  initStageClearAtr();
}

static void move() {
  switch ( status ) {
  case TITLE:
    moveTitleMenu();
    moveBackground();
    addBullets();
    moveFoes();
    break;
  case IN_GAME:
    moveBackground();
    addBullets();
    moveShots();
    moveShip();
    moveFoes();
    moveFrags();
    moveBonuses();
    break;
  case GAMEOVER:
    moveGameover();
    moveBackground();
    addBullets();
    moveShots();
    moveFoes();
    moveFrags();
    break;
  case STAGE_CLEAR:
    moveStageClear();
    moveBackground();
    moveShots();
    moveShip();
    moveFrags();
    moveBonuses();
    break;
  case PAUSE:
    movePause();
    break;
  }
}

static void draw() {
  switch ( status ) {
  case TITLE:
    // Draw background.
    drawBackground();
    drawFoes();
    drawBulletsWake();
    blendScreen();
    // Draw forground.
    drawBullets();
    drawScore();
    drawTitleMenu();
    break;
  case IN_GAME:
    // Draw background.
    drawBackground();
    drawBonuses();
    drawFoes();
    drawBulletsWake();
    drawFrags();
    blendScreen();
    // Draw forground.
    drawShots();
    drawShip();
    drawBullets();
    drawScore();
    break;
  case GAMEOVER:
    // Draw background.
    drawBackground();
    drawFoes();
    drawBulletsWake();
    drawFrags();
    blendScreen();
    // Draw forground.
    drawShots();
    drawBullets();
    drawScore();
    drawGameover();
    break;
  case STAGE_CLEAR:
    // Draw background.
    drawBackground();
    drawBonuses();
    drawFrags();
    blendScreen();
    // Draw forground.
    drawShots();
    drawShip();
    drawScore();
    drawStageClear();
    break;
  case PAUSE:
    // Draw background.
    drawBackground();
    drawBonuses();
    drawFoes();
    drawBulletsWake();
    drawFrags();
    blendScreen();
    // Draw forground.
    drawShots();
    drawShip();
    drawBullets();
    drawScore();
    drawPause();
    break;
  }
}


static int accframe = 0;

static void usage(char *argv0) {
  fprintf(stderr, "Usage: %s [-nosound] [-window] [-reverse] [-brightness n] [-nowait] [-accframe]\n", argv0);
}

static void parseArgs(int argc, char *argv[]) {
  int i;
  for ( i=1 ; i<argc ; i++ ) {
    if ( strcmp(argv[i], "-nosound") == 0 ) {
      noSound = 1;
    } else if ( strcmp(argv[i], "-window") == 0 ) {
      windowMode = 1;
    } else if ( strcmp(argv[i], "-reverse") == 0 ) {
      buttonReversed = 1;
    } else if ( (strcmp(argv[i], "-brightness") == 0) && argv[i+1] ) {
      i++;
      brightness = (int)atoi(argv[i]);
      if ( brightness < 0 || brightness > 256 ) {
	brightness = DEFAULT_BRIGHTNESS;
      }
    } else if ( strcmp(argv[i], "-nowait") == 0 ) {
      nowait = 1;
    } else if ( strcmp(argv[i], "-accframe") == 0 ) {
      accframe = 1;
    } else {
      usage(argv[0]);
      exit(1);
    }
  }
}

int interval = INTERVAL_BASE;
int tick = 0;
static int pPrsd = 1;

int main(int argc, char *argv[]) {
  int done = 0;
  long prvTickCount = 0;
  int i;
  SDL_Event event;
  long nowTick;
  int frame;

  parseArgs(argc, argv);

  initDegutil();
  initSDL(windowMode);
  if ( !noSound ) initSound();
  initFirst();
  initTitle();

  while ( !done ) {
    int startBtn = 0, backBtn = 0;
    SDL_PollEvent(&event);
    keys = SDL_GetKeyState(NULL);
    startBtn = SDL_GameControllerGetButton(gamepad, SDL_CONTROLLER_BUTTON_START);
    backBtn = SDL_GameControllerGetButton(gamepad, SDL_CONTROLLER_BUTTON_BACK);
    if ( keys[SDLK_ESCAPE] == SDL_PRESSED || event.type == SDL_QUIT || backBtn ) done = 1;
    if ( keys[SDLK_p] == SDL_PRESSED || startBtn ) {
      if ( !pPrsd ) {
	if ( status == IN_GAME ) {
	  status = PAUSE;
	} else if ( status == PAUSE ) {
	  status = IN_GAME;
	}
      }
      pPrsd = 1;
    } else {
      pPrsd = 0;
    }

    nowTick = SDL_GetTicks();
    frame = (int)(nowTick-prvTickCount) / interval;
    if ( frame <= 0 ) {
      frame = 1;
      SDL_Delay(prvTickCount+interval-nowTick);
      if ( accframe ) {
	prvTickCount = SDL_GetTicks();
      } else {
	prvTickCount += interval;
      }
    } else if ( frame > 5 ) {
      frame = 5;
      prvTickCount = nowTick;
    } else {
      prvTickCount += frame*interval;
    }
    for ( i=0 ; i<frame ; i++ ) {
      move();
      tick++;
    }
    smokeScreen();
    draw();
    flipScreen();
  }
  quitLast();
}
